python/tvm/topi/arm_cpu/mprofile/dsp/tensordot_conv2ds.py

```diff
@@ -0,0 +1,276 @@
+    improves accuracy on models imported from TensorFlow."""
+    return (data_dim - kernel_dim + pad_before + pad_after) % stride
+
+
+def _get_suffix() -> str:
+    """Returns a random eight-character string to append to C function names. Prevents accidental
+    re-definition of functions if the same operator appears twice in a Relay graph."""
+    return "".join(random.choices(string.ascii_uppercase, k=8))
+
+
+def conv2d_nhwc_ohwi_dsp_compute(_cfg, data, kernel, strides, padding, dilation, out_dtype):
+    """Standard conv2d schedule that can be tensorized using tensordot."""
+
+    stride_h, stride_w = _unpack_2d_argument(strides)
+    pad_up, pad_left, pad_down, pad_right = _unpack_padding(padding)
+    _check_no_dilation(dilation)
+
+    batch_size, data_h, data_w, in_channels = data.shape
+    output_channels, kernel_h, kernel_w, _ = kernel.shape
+    assert kernel.shape[3] == in_channels
+
+    output_h = _compute_output_dim(data_h, kernel_h, pad_up, pad_down, stride_h)
+    output_w = _compute_output_dim(data_w, kernel_w, pad_left, pad_right, stride_w)
+    y_offset = _compute_offset(data_h, kernel_h, pad_up, pad_down, stride_h)
+    x_offset = _compute_offset(data_w, kernel_w, pad_left, pad_right, stride_w)
+
+    kh_i = te.reduce_axis((0, kernel_h), name="kh_i")
+    kw_i = te.reduce_axis((0, kernel_w), name="kw_i")
+    kc_i = te.reduce_axis((0, in_channels), name="rc")
+
+    padded_data = _pad_if_needed(data, "NHWC", (pad_up, pad_left, pad_down, pad_right))
+    return te.compute(
+        (batch_size, output_h, output_w, output_channels),
+        lambda n, y, x, c: te.sum(
+            padded_data[
+                n, y_offset + y * stride_h + kh_i, x_offset + x * stride_w + kw_i, kc_i
+            ].astype(out_dtype)
+            * kernel[c, kh_i, kw_i, kc_i].astype(out_dtype),
+            axis=(kh_i, kw_i, kc_i),
+        ),
+        name="conv2d",
+        tag="conv2d_nhwc_ohwi_dsp",
+    )
+
+
+def _make_conv2d_tensorization(padded_data, kernel):
+    _, _, padded_w, in_channels = padded_data.shape
+    _, kernel_h, kernel_w, _ = kernel.shape
+    in_dtype = padded_data.dtype
+    suffix = _get_suffix()
+    assert in_dtype == kernel.dtype
+
+    data_slice = te.placeholder((kernel_h, kernel_w, in_channels), name="a", dtype=in_dtype)
+    kernel_slice = te.placeholder((kernel_h, kernel_w, in_channels), name="b", dtype=in_dtype)
+
+    kh_i = te.reduce_axis((0, kernel_h), name="kh_i")
+    kw_i = te.reduce_axis((0, kernel_w), name="kw_i")
+    kc_i = te.reduce_axis((0, in_channels), name="kc_i")
+
+    output_slice = te.compute(
+        (1,),
+        lambda k: te.sum(
+            data_slice[kh_i, kw_i, kc_i].astype("int32")
+            * kernel_slice[kh_i, kw_i, kc_i].astype("int32"),
+            axis=[kh_i, kw_i, kc_i],
+        ),
+        name="c",
+    )
+
+    # TVM has a really strange bug where the outer reduction axis (kh_i) having length 1 causes the
```

Comment:
did you file an issue for this? cc @vinx13

Reply:
I've filed #13010 to track this. I also [wrote up a Colab notebook](https://colab.research.google.com/drive/1Y9LXBdQxQD-FjNbW6cChuExse6IvMdGZ?usp=sharing) to allow the issue to be easily reproduced.